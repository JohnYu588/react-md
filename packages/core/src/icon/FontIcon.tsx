import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";

import type { FontIconClassNameOptions } from "./styles";
import { icon } from "./styles";

export interface FontIconProps
  extends Omit<HTMLAttributes<HTMLElement>, "color">,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 *
 * \@remarks \@since 6.0.0 Switched from `<i>` to `<span>` element.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(props, ref) {
    const {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      color,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...remaining
    } = props;

    return (
      <span
        {...remaining}
        aria-hidden={ariaHidden}
        ref={ref}
        className={icon({
          type: "font",
          dense,
          color,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </span>
    );
  }
);
