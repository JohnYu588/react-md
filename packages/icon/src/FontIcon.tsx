import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";
import type { FontIconClassNameOptions } from "./getIconClassName";
import { getIconClassName } from "./getIconClassName";

export interface FontIconProps
  extends HTMLAttributes<HTMLElement>,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(
    {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...props
    },
    ref
  ) {
    return (
      <i
        {...props}
        aria-hidden={ariaHidden}
        ref={ref}
        className={getIconClassName({
          type: "font",
          dense,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </i>
    );
  }
);
