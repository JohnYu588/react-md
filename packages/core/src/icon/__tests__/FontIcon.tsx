import { describe, expect, it } from "@jest/globals";

import { render, screen } from "../../test-utils/index.js";
import { FontIcon } from "../FontIcon.js";

describe("FontIcon", () => {
  it("should render as an <i> element", () => {
    render(<FontIcon data-testid="icon">home</FontIcon>);
    expect(screen.getByTestId("icon")).toBeInstanceOf(HTMLSpanElement);
  });

  it("should render correctly", () => {
    const { rerender } = render(<FontIcon data-testid="icon">home</FontIcon>);
    const icon = screen.getByTestId("icon");
    expect(icon).toMatchSnapshot();

    rerender(<FontIcon data-testid="icon" iconClassName="fa fa-github" />);
  });

  it("should apply the correct classNames", () => {
    const { rerender } = render(<FontIcon data-testid="icon" />);
    const icon = screen.getByTestId("icon");
    expect(icon).toHaveClass("rmd-icon rmd-icon--font material-icons", {
      exact: true,
    });
    expect(icon).toMatchSnapshot();

    rerender(<FontIcon data-testid="icon" className="test-thing" dense />);
    expect(icon).toHaveClass(
      "rmd-icon rmd-icon--font rmd-icon--dense material-icons test-thing",
      { exact: true }
    );
    expect(icon).toMatchSnapshot();

    rerender(
      <FontIcon
        data-testid="icon"
        className="test-thing"
        dense
        iconClassName="fa fa-github"
      />
    );
    expect(icon).toHaveClass(
      "rmd-icon rmd-icon--font rmd-icon--dense fa fa-github test-thing",
      { exact: true }
    );
    expect(icon).toMatchSnapshot();
  });
});
