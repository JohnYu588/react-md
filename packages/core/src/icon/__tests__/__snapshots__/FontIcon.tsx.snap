// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontIcon should apply the correct classNames 1`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--font material-icons"
  data-testid="icon"
/>
`;

exports[`FontIcon should apply the correct classNames 2`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--font rmd-icon--dense material-icons test-thing"
  data-testid="icon"
/>
`;

exports[`FontIcon should apply the correct classNames 3`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--font rmd-icon--dense fa fa-github test-thing"
  data-testid="icon"
/>
`;

exports[`FontIcon should render correctly 1`] = `
<span
  aria-hidden="true"
  class="rmd-icon rmd-icon--font material-icons"
  data-testid="icon"
>
  home
</span>
`;
